lib/propolis/src/block/mod.rs

```diff
@@ -224,6 +225,13 @@ pub trait Device: Send + Sync + 'static {
     /// Complete processing of result
     fn complete(&self, res: Result, id: ReqId);
 
+    /// Attach a callback to be run on completion of I/Os.
+    ///
+    /// Returns whether there was a previously-registered callback.
+    fn on_completion(&self, _cb: Box<dyn CompletionCallback>) -> bool {
+        false
```

Comment:
I'm not wild about there being a default implementation here.  It would be misleading for a block device to accept setting a completion callback, only to never call it for any of its completed requests.

Reply:
That's fair. How about returning something like an `anyhow::Result<bool>`? That would return whether the callback existed previously in the `Ok(_)` variant, or fail if the device chose not to support it.

Reply:
> if the device chose not to support it.

All of the existing devices do support it, so I wouldn't worry about that sort of thing for now.  If things change in the future, we can deal with it then.

Reply:
So to be concrete, you're only arguing that we remove the default impl? Otherwise everything else remains unchanged about this interface?

Reply:
> you're only arguing that we remove the default impl

yeah, the rest is fine as-is

Reply:
Removed the default impl in 8565b0c